src/main/scala/tech/cryptonomic/conseil/tezos/TezosNodeOperator.scala

```diff
@@ -103,158 +107,170 @@ class TezosNodeOperator(node: TezosRPCInterface) extends LazyLogging {
-    * @param followFork     If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
-    * @return               Blocks
+    * Fetches metadata for a block only, without waiting for the answer
+    * @param network   Which Tezos network to go against
+    * @param hash      Hash of the block
+    * @return          the metadata wrapped in a [[Future]]
     */
-  def getBlocks(network: String, offset: Int, startBlockHash: Option[String], followFork: Boolean): Try[List[Block]] =
-    startBlockHash match {
-      case None =>
-        getBlockHead(network).flatMap(head => getBlocks(network, head.metadata.header.level - offset + 1, head.metadata.header.level, startBlockHash, followFork))
-      case Some(hash) =>
-        getBlock(network, hash).flatMap(block => getBlocks(network, block.metadata.header.level - offset + 1, block.metadata.header.level, startBlockHash, followFork))
-    }
+  private def getMetadata(network: String, hash: String): Future[BlockMetadata] =
+    node.runAsyncGetQuery(network, s"blocks/$hash") map fromJson[BlockMetadata]
 
   /**
-    * Gets the blocks using a specified rage
-    * @param network        Which Tezos network to go against
-    * @param minLevel       Minimum block level
-    * @param maxLevel       Maximum block level
-    * @param startBlockHash If specified, start from the supplied block hash.
-    * @param followFork     If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
-    * @return               Blocks
+    * Asynchronously fetches a whole chain of metadata for blocks, using the parameters to
+    * decide when to stop the recursive descent on predecessors
+    *
+    * @param network    which Tezos network to go against
+    * @param current    The current metadata object to start descending
+    * @param minLevel   The lowest level for which to include results, unless a fork should be followed
+    * @param maxLevel   The highest level for which to include results
+    * @param followFork If the predecessor of the minLevel block appears to be on a fork, also capture the blocks on the fork.
+    * @param collected  The currently accumulated metadata so far
+    * @return           The metadata list, wrapped in a [[Future]]
     */
-  def getBlocks(network: String, minLevel: Int, maxLevel: Int, startBlockHash: Option[String], followFork: Boolean): Try[List[Block]] =
-    startBlockHash match {
-      case None =>
-        getBlockHead(network).flatMap(head => Try {
-          processBlocks(network, head.metadata.hash, minLevel, maxLevel, followFork = followFork)
-        })
-      case Some(hash) =>
-        getBlock(network, hash).flatMap(block => Try {
-          processBlocks(network, block.metadata.hash, minLevel, maxLevel, followFork = followFork)
-        })
-    }
+  def recursiveMetadataFetch(
```

Comment:
okay, this part should definitely be scrapped. I've been working on a rewrite for this function that's significantly more intuitive (once it's debugged). This requires some domain knowledge, it won't be easily fixed

Reply:
actually, everything after this line should either be moved to a separate PR, or scrapped completely.

Reply:
The upshot of doing it this way is we compare strictly sync vs. async logic. Once we convince ourselves the async logic is correct and performant, we can add the improved block fetching functionality to the async code.

Reply:
@ba205 I perfectly understand your point, this part is the result from my need to load data from zeronet without taking so long, as I needed to compare data for the other work I's on.

As I said elsewhere, I don't expect this code to be merged.

Yet I took the chance to put the code here for review, because:
 1. I wanted to check with you both (which definitely have more domain knowledge!) if my assumptions on simplifying were generally correct (this will be of huge help to make me learn)
 2. I wanted to check with you how an async approach would look like and verify what needs to be fixed to make it correct and be able to later adopt a similar design on the real block-processing rewrite

I beg your patience here if you feel this is wasted time, but I think this work is useful to define the final shape that the code will eventually take